### PR TITLE
[Merged by Bors] - Concurrent validation

### PR DESCRIPTION
**Pull request by almogdepaz**
## Motivation
allows us to validate blocks faster during sync

## Changes
validation queue to handle each block concurrently 

## Test Plan
same as before

### CODE REVIEW

File: p2p/net/network.go [hunk 1]
```diff
@@ -181,8 +181,8 @@ func dial(keepAlive, timeOut time.Duration, address net.Addr) (net.Conn, error)
 func tcpSocketConfig(tcpconn *net.TCPConn) {
 	// TODO: Error handling, what if only certain flags are supported
 	// TODO: Parameters, try to find right buffers based on something or os/net-interface input?
-	tcpconn.SetReadBuffer(1024 * 64)
-	tcpconn.SetWriteBuffer(1024 * 64)
+	tcpconn.SetReadBuffer(5 * 1024 * 1024)
```
Comment (gavraz) [inline p2p/net/network.go:184]: maybe extract const MB = 1024*1024

File: activation/activation.go [hunk 2]
```diff
@@ -14,7 +14,7 @@ import (
 
 const AtxProtocol = "AtxGossip"
 
-var activesetCache = NewActivesetCache(1000)
+var activesetCache = NewActivesetCache(100)
```
Comment (gavraz) [inline activation/activation.go]: Better leave as it was.
We don't want to play with these numbers right now. We know it works with 1000. 100 might be too low.

File: sync/validation_queue.go [hunk 3]
```diff
@@ -64,33 +64,37 @@ func (vq *blockQueue) inQueue(id types.Hash32) bool {
...
-		block, found := mp[id]
-		if !found {
+		b, ok := mp[id]
+		if !ok {
 			vq.updateDependencies(id, false)
 			vq.Error("could not retrieve a block in view %v", id.ShortString())
 			continue
 		}
 
+		go vq.handleBlock(id, b)
```
Comment (gavraz) [inline sync/validation_queue.go:82]: How big can bjb.ids be?
We probably don't want to open a go-routine for each block if this number is very very big.

File: sync/validation_queue.go [hunk 4]
```diff
@@ -64,33 +64,37 @@ func (vq *blockQueue) inQueue(id types.Hash32) bool {
...
+	}
+
+}
+
+func (vq *blockQueue) handleBlock(id types.Hash32, b *types.Block) {
+	func(block *types.Block) {
```
Comment (gavraz) [inline sync/validation_queue.go]: why call this inner function? you just create a wrapping function and call it...

**Review by gavraz — CHANGES_REQUESTED** [review]

File: activation/activation.go (hunk 2 above)
Comment (almogdepaz) [reply activation/activation.go]: @barakshani please tell @gavraz what params you are actually using

File: sync/validation_queue.go (hunk 3 above)
Comment (almogdepaz) [reply sync/validation_queue.go:82]: depends on batch sizes right now 50, anyway its along those lines. this is the same behavior as already implemented during gossip as we also handle gossip blocks concurrently.

File: sync/validation_queue.go (hunk 4 above)
Comment (almogdepaz) [reply sync/validation_queue.go]: auto refactor tool, i will fix it

File: activation/activation.go (hunk 2 above)
Comment (barakshani) [reply activation/activation.go]: I ran with 100, but we can stick with 1000. The items in this cache are very small (hash,integer).

Comment (almogdepaz) [reply activation/activation.go]: tov

**Review by gavraz — APPROVED** [review]

**Comment by almogdepaz** [comment]
bors merge

**Pull request closed without merging**